source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -503,14 +503,18 @@ void DnsMessageParser::setResponseCode(DnsQueryContextPtr& context,
   context->response_code_ = DNS_RESPONSE_CODE_NO_ERROR;
 }
 
-void DnsMessageParser::generateRandomIndices(const size_t count,
-                                             absl::flat_hash_set<size_t>& elements) {
-  elements.clear();
-  elements.reserve(count);
-  while (elements.size() < count) {
-    size_t element = rng_.random() % count;
-    if (!elements.contains(element)) {
-      elements.emplace(element);
+void DnsMessageParser::generateRandomIndices(absl::FixedArray<size_t>& index_array) {
+  const size_t indices = index_array.size();
+
+  absl::flat_hash_set<size_t> used_indices{};
```

Comment:
Sorry, I didn't notice that you are using the set to dedup the returned values. That makes sense. I would revert what you had before, however, I would recommend optimizing so that you don't build the set in the case where all the records will fit (you should be able to compute this I think).

/wait

Reply:
The set is constructed before we check the buffer size limits.  This is just randomizing the order that we choose the answer records.

I know Route53 imposes an 8 record limit in an answer, perhaps we can limit the number of answers to 8 or less to reduce how much work is done here.